ehr/src/org/labkey/ehr/utils/TriggerScriptHelper.java

```diff
@@ -1215,30 +1223,128 @@ private String checkOtherDrawsQuantity(String id, Date date, String rowObjectId,
             {
                 overages.add(bloodNextInterval);
             }
+            else if (doWarnForBloodNearOverages())
+            {
+                //only report about problematic bloods nearing the limit in the current transaction
+                //because allBloods contains everything and we want to distinguish them
+                if (blood1.getInTransaction())
+                {
+                    double maxAllowableThreshold = maxAllowable - getBloodNearingOveragesThreshold();
+                    if (bloodNextInterval > maxAllowableThreshold)
+                    {
+                        closeToThreshold.add(bloodNextInterval);
+                    }
+                }
+            }
         }
 
-        //always report the most severe overage
-        if (!overages.isEmpty())
+        if (!overages.isEmpty() || !closeToThreshold.isEmpty())
         {
-            return "Blood volume of " + rowQuantity + " (" + overages.descendingSet().iterator().next() + " over " + interval + " days) exceeds the allowable volume of " + maxAllowable + " mL (weight: " + weight + " kg)";
+            StringBuilder errorMsgBuilder = new StringBuilder();
+
+            //always report the most severe overage
+            if (!overages.isEmpty())
+            {
+                errorMsgBuilder.append("Blood volume of ")
+                               .append(rowQuantity)
+                               .append(" (")
+                               .append(overages.descendingSet().iterator().next())
+                               .append(" over ")
+                               .append(interval)
+                               .append(" days) exceeds the allowable volume of ")
+                               .append(maxAllowable)
+                               .append(" mL (weight: ")
+                               .append(weight)
+                               .append(" kg).\n");
+            }
+
+            //report all the ones that are close to limit
+            for (double closeValue : closeToThreshold.descendingSet())
+            {
+                errorMsgBuilder.append("Limit notice! Blood volume of ")
```

Comment:
@csebranek: i realize this comment was closed. I question whether there is any point in looping over the list of near-overages. It is possible for a given blood draw to have multiple overages and/or multiple near-overages, since it could overlap with varying numbers of future records. With overages, the code just reports the max(overage), rather than giving one message for every overage.

Is there any reason you shouldnt to the exact same thing here? In what instance would having multiple nearly identical messages be preferable to reporting the max(closeValue)?

Reply:
I see your point, not much use in having error messages representing nearly the same information, i distilled it down to one